qiskit/algorithms/fidelities/base_fidelity.py

```diff
@@ -0,0 +1,253 @@
+    (parametrized) quantum circuits.
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        """Initializes the class to evaluate fidelities."""
+
+        self._circuits: Sequence[QuantumCircuit] = []
+        self._parameter_values: Sequence[Sequence[float]] = []
+
+        # use cache for preventing unnecessary circuit compositions
+        self._circuit_cache: Mapping[(int, int), QuantumCircuit] = {}
+
+        self._left_parameters = []
+        self._right_parameters = []
+
+    def _check_values_shape(
+        self,
+        values: Sequence[Sequence[float]] | None,
+        circuits: QuantumCircuit,
+        label: str | None = " ",
+    ) -> list[list[float]] | None:
+        """
+        Checks whether the passed values match the shape of the parameters
+        of the corresponding circuits and formats values to 2D list.
+
+        Args:
+            values: parameter values corresponding to the circuits to be checked
+            circuits: list of circuits to be checked
+            label: optional label to allow for circuit identification in error message
+
+        Returns:
+            Returns a 2D list if values match, `None` if no parameters are passed
+
+        Raises:
+            ValueError: if the number of parameter values doesn't match the number of
+                        circuit parameters
+        """
+
+        # Support ndarray
+        if isinstance(values, np.ndarray):
+            values = values.tolist()
+
+        if values is None:
+            for circuit in circuits:
+                if circuit.num_parameters != 0:
+                    raise ValueError(
+                        f"`values_{label}` cannot be `None` because circuit_{label} has "
+                        f"{circuit.num_parameters} free parameters."
+                    )
+            return None
+        else:
+            # ensure 2d list
+            if not isinstance(values, list):
+                values = [values]
+            if len(values) > 0 and not isinstance(values[0], list):
+                values = [values]
+            return values
+
+    def _check_qubits_match(self, circuit_1: QuantumCircuit, circuit_2: QuantumCircuit) -> None:
+        """
+        Checks that the number of qubits of 2 circuits matches.
+        Args:
+            circuit_1: (Parametrized) quantum circuit
+            circuit_2: (Parametrized) quantum circuit
+
+        Raises:
+            ValueError: when ``circuit_1`` and ``circuit_2`` don't have the same number of qubits.
+        """
+
+        if circuit_1 is not None and circuit_2 is not None:
+            if circuit_1.num_qubits != circuit_2.num_qubits:
+                raise ValueError(
+                    f"The number of qubits for the left circuit ({circuit_1.num_qubits}) "
+                    f"and right circuit ({circuit_2.num_qubits}) do not coincide."
+                )
+
+    @abstractmethod
+    def _create_fidelity_circuit(
```

Comment:
We could even make this method public, as `construct_circuit` it could be useful if users want to play around with the fidelities 🙂

Reply:
I am not super convinced about making it public because it would mess up with the circuit cache...